pkg/k8s/watchers/crd.go

```diff
@@ -0,0 +1,294 @@
+// limitations under the License.
+
+package watchers
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	"github.com/cilium/cilium/pkg/k8s/informer"
+	k8sversion "github.com/cilium/cilium/pkg/k8s/version"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/option"
+
+	apiextclientset "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+)
+
+func (k *K8sWatcher) crdsInit(client apiextclientset.Interface) {
+	crds := newCRDState(
+		map[string]bool{
+			cnpCRD:  false,
+			ccnpCRD: false,
+			cepCRD:  false,
+			cnCRD:   false,
+			cidCRD:  false,
+			clrpCRD: false,
+		},
+	)
+
+	_, crdController := informer.NewInformer(
+		newListWatchFromClient(
+			newCRDGetter(client),
+			fields.Everything(),
+		),
+		&metav1.PartialObjectMetadata{},
+		0,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				if pom, ok := obj.(*metav1.PartialObjectMetadata); ok {
+					crds.Lock()
+					crds.m["crd:"+pom.GetName()] = true
+					crds.Unlock()
+				}
+			},
+			DeleteFunc: func(obj interface{}) {
+				if pom, ok := obj.(*metav1.PartialObjectMetadata); ok {
+					crds.Lock()
+					crds.m["crd:"+pom.GetName()] = false
+					crds.Unlock()
+				}
+			},
+		},
+		nil,
+	)
+
+	// Create a context so that we can timeout after the configured CRD wait
+	// peroid.
+	ctx, cancel := context.WithTimeout(context.Background(), option.Config.CRDWaitTimeout)
+	for crd := range crds.m {
+		k.blockWaitGroupToSyncResources(
+			ctx.Done(),
+			nil,
+			func() bool {
+				crds.Lock()
+				defer crds.Unlock()
+				return crds.m[crd]
+			},
+			crd,
+		)
+	}
+
+	// The above loop will call blockWaitGroupToSyncResources to populate the
+	// K8sWatcher state with the current state of the CRDs. It will check the
+	// state of each CRD, with the function provided below. If the function
+	// reports that the given CRD is true (has been synced), it will close a
+	// channel associated with the given CRD. A subsequent call to
+	// (*K8sWatcher).WaitForCacheSync will notice that a given CRD's channel
+	// has been closed. Once all the CRDs passed to WaitForCacheSync have had
+	// their channels closed, the function unblocks.
+	//
+	// Meanwhile, the Goroutine below kicks off the controller that was
+	// instantiated above, and enters a loop looking for (1) if the context has
+	// deadlined or (2) if the entire CRD state has been synced (all CRDs found
+	// in the cluster). If (1) is encountered, then Cilium will fatal because
+	// it cannot proceed if the CRDs are not present. If (2) is encountered,
+	// then make sure the controller has exited by cancelling the context and
+	// we return out.
```

Comment:
```suggestion
	// Meanwhile, the Goroutine below kicks off the controller that was
	// instantiated above, and enters a loop looking for (1) if the context has
	// deadlined or (2) if the entire CRD state has been synced (all CRDs found
	// in the cluster). While the Goroutine is in this loop, the controller is listening
	// for either add or delete events to the customresourcedefinition resource 
	// (disguised inside a metav1.PartialObjectMetadata object).
	// If (1) is encountered, then Cilium will fatal because
	// it cannot proceed if the CRDs are not present. If (2) is encountered,
	// then make sure the controller has exited by cancelling the context and
	// we return out.
```